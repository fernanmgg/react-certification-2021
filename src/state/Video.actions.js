export const actions = (dispatch) => ({
  setSearch: (payload) => dispatch({ type: 'SET_SEARCH', payload }),
  setAuth: (payload) => dispatch({ type: 'SET_AUTH', payload }),
  unsetAuth: () => dispatch({ type: 'UNSET_AUTH' }),
  addFav: (payload) => dispatch({ type: 'ADD_FAVORITE', payload }),
  remFav: (payload) => dispatch({ type: 'REMOVE_FAVORITE', payload }),
});
