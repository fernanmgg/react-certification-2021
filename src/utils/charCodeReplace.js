function charCodeReplace(input) {
  return input
    .replace(/&#(\d+);/g, (_, code) => {
      return String.fromCharCode(parseInt(code, 10));
    })
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"');
}

export default charCodeReplace;
