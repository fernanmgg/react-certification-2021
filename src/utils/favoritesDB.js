function getFavorites(name) {
  const localData = localStorage.getItem(name);
  if (localData) {
    return JSON.parse(localData);
  }
  return [];
}

function isFavorite(name, videoId) {
  const localData = localStorage.getItem(name);
  if (localData) {
    return JSON.parse(localData).some((element) => element.id.videoId === videoId);
  }
  return false;
}

function addFavorite(name, video) {
  const localData = localStorage.getItem(name);
  if (localData) {
    localStorage.setItem(name, JSON.stringify([video, ...JSON.parse(localData)]));
  } else {
    localStorage.setItem(name, JSON.stringify([video]));
  }
}

function removeFavorite(name, videoId) {
  const localData = localStorage.getItem(name);
  if (localData) {
    const videos = JSON.parse(localData).filter(
      (element) => element.id.videoId !== videoId
    );
    localStorage.setItem(name, JSON.stringify(videos));
  }
}

export { getFavorites, isFavorite, addFavorite, removeFavorite };
