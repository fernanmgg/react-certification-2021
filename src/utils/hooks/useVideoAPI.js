import { useState, useEffect } from 'react';

const key = process.env.REACT_APP_YOUTUBE_API_KEY;

const baseUrl = 'https://www.googleapis.com/youtube/v3/videos?';

export default function useVideoAPI(videoId) {
  const [video, setVideo] = useState(null);
  const [redirect, setRedirect] = useState(false);
  const [fetchError, setFetchError] = useState(false);

  useEffect(() => {
    function buildURI() {
      const fields = [
        `id=${videoId}`,
        'part=snippet',
        'fields=items(snippet(title,description,thumbnails(medium(url))))',
        `key=${key}`,
      ];
      return baseUrl.concat(
        fields.reduce((previous, current) => previous.concat(`&${current}`))
      );
    }

    const fetchData = async () => {
      try {
        const response = await fetch(encodeURI(buildURI()));
        const json = await response.json();
        if (Array.isArray(json.items) && json.items.length === 1) setVideo(json.items[0]);
        else setRedirect(true);
      } catch (e) {
        setFetchError(true);
      }
    };

    const check = /^[A-Za-z0-9_-]{11}$/.test(videoId);
    if (check) fetchData();
    else setRedirect(true);
  }, [videoId]);

  return { video, redirect, fetchError };
}
