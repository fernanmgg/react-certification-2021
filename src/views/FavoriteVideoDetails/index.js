export { default } from './FavoriteVideoDetails.view';
