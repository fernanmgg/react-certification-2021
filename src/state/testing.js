import React from 'react';
import { ThemeProvider } from 'styled-components';

import { VideoContext } from './Video.state';
import { lightTheme as theme } from '../components/App/App.theme';

const wrapWithThemeContext = (children) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const wrapWithVideoContext = (children, state = {}, actions = {}) => {
  return (
    <VideoContext.Provider value={{ state, ...actions }}>
      {children}
    </VideoContext.Provider>
  );
};

export { wrapWithThemeContext, wrapWithVideoContext };
