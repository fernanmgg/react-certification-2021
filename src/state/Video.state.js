import React from 'react';

import { getFavorites } from '../utils/favoritesDB';

const VideoContext = React.createContext();

const initialState = {
  auth: null,
  favorites: [],
  search: 'wizeline',
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_SEARCH':
      return { ...state, search: action.payload };
    case 'SET_AUTH':
      if (action.payload.remember)
        localStorage.setItem('auth', JSON.stringify(action.payload.auth));
      else sessionStorage.setItem('auth', JSON.stringify(action.payload.auth));
      return {
        ...state,
        auth: action.payload.auth,
        favorites: getFavorites(action.payload.auth.name),
      };
    case 'UNSET_AUTH':
      localStorage.removeItem('auth');
      sessionStorage.removeItem('auth');
      return { ...state, auth: null, favorites: [] };
    default:
      throw new Error('Error: Undefined action');
  }
};

export { VideoContext, initialState, reducer };
