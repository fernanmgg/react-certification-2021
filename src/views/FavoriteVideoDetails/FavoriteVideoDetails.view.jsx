import React, { useContext } from 'react';

import Details from '../../components/Details';
import { VideoContext } from '../../state/Video.state';
import { getFavoritesInfo } from '../../utils/favoritesDB';

function FavoriteVideoDetails() {
  const { state } = useContext(VideoContext);
  const { auth, favorites } = state;
  return <Details videos={getFavoritesInfo(auth.id, favorites)} />;
}

export default FavoriteVideoDetails;
