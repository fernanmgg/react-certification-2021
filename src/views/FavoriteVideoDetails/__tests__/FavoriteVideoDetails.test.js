import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router';
import user from '@testing-library/user-event';

import FavoriteVideoDetails from '../FavoriteVideoDetails.view';
import * as useVideoAPI from '../../../utils/hooks/useVideoAPI';
import { wrapWithVideoContext } from '../../../state/testing';
import * as favoritesDB from '../../../utils/favoritesDB';

const video = {
  snippet: {
    title: 'test title',
    description: 'test description',
    thumbnails: {
      medium: {
        url: 'test',
      },
    },
  },
};

useVideoAPI.default = jest.fn(() => ({ video }));

describe('FavoriteVideoDetails UI tests', () => {
  test('renders details with correct props', () => {
    render(
      wrapWithVideoContext(
        <MemoryRouter initialEntries={['/favorites/Test id']}>
          <Route path="/favorites/:videoId">
            <FavoriteVideoDetails />
          </Route>
        </MemoryRouter>,
        { auth: { name: 'test' }, favorites: [] },
        jest.fn()
      )
    );
    const title = screen.getByText(/test title/i);
    const description = screen.getByText(/test description/i);
    expect(title).toBeInTheDocument();
    expect(description).toBeInTheDocument();
    expect(useVideoAPI.default).toHaveBeenCalled();
    expect(useVideoAPI.default).toHaveBeenCalledWith('Test id');
  });

  test('renders "Add favorite" button if the video is not favorite', () => {
    const dispatch = jest.fn();
    favoritesDB.isFavorite = jest.fn(() => false);
    favoritesDB.addFavorite = jest.fn();
    render(
      wrapWithVideoContext(
        <MemoryRouter initialEntries={['/video/testId']}>
          <Route path="/video/:videoId">
            <FavoriteVideoDetails />
          </Route>
        </MemoryRouter>,
        { auth: { name: 'test' }, favorites: [] },
        dispatch
      )
    );
    const favorite = screen.getByRole('button', { name: /add favorite/i });
    expect(favorite).toBeInTheDocument();
    user.click(favorite);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: 'ADD_FAVORITE', payload: 'testId' });
  });

  test('renders "Remove favorite" button if the video is favorite', () => {
    const dispatch = jest.fn();
    favoritesDB.isFavorite = jest.fn(() => true);
    favoritesDB.removeFavorite = jest.fn();
    render(
      wrapWithVideoContext(
        <MemoryRouter initialEntries={['/video/testId']}>
          <Route path="/video/:videoId">
            <FavoriteVideoDetails />
          </Route>
        </MemoryRouter>,
        { auth: { name: 'test' }, favorites: [] },
        dispatch
      )
    );
    const favorite = screen.getByRole('button', { name: /remove favorite/i });
    expect(favorite).toBeInTheDocument();
    user.click(favorite);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ type: 'REMOVE_FAVORITE', payload: 'testId' });
  });
});
