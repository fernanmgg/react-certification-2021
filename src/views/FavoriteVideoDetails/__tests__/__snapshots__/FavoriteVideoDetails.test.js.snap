// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteVideoDetails UI tests renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: calc(100vh - 64px - 16px);
  margin-top: 16px;
  padding: 16px;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 3 1 0;
  -ms-flex: 3 1 0;
  flex: 3 1 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  margin: 0 16px;
}

.c2 {
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  margin-top: 8px;
}

.c5 {
  border: 1px solid;
  border-radius: 8px;
  cursor: pointer;
  font-size: 1rem;
  margin: 0 8px;
  padding: 8px;
}

.c6 {
  font-size: 1.5rem;
}

.c7 {
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  font-size: 1rem;
  overflow-y: auto;
  white-space: pre-wrap;
}

.c8 {
  border-radius: 8px;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  overflow-y: scroll;
  padding-top: 8px;
}

.c9 {
  text-align: center;
  width: 100%;
}

@media (max-width:900px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    height: 200vh;
  }
}

@media (max-width:900px) {
  .c1 {
    margin: 16px 0;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <iframe
          height="100%"
          src="https://www.youtube.com/embed/testId"
          title="testId"
          width="100%"
        />
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <button
            class="c5"
          >
            Add favorite
          </button>
        </div>
        <div
          class="c6"
        >
          test title
        </div>
        <div
          class="c7"
        >
          test description
        </div>
      </div>
    </div>
    <div
      class="c8"
    >
      <h2
        class="c9"
      >
        No videos found...
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
